WebDriverAgentLib/Utilities/FBTVNavigationTracker.m

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) 2018-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "FBTVNavigationTracker.h"
+
+#import "FBApplication.h"
+#import "FBMathUtils.h"
+#import "XCUIApplication+FBFocused.h"
+#import "XCUIElement+FBUtilities.h"
+#import "XCUIElement+FBWebDriverAttributes.h"
+
+@interface FBTVNavigationItem : NSObject
+@property (nonatomic, assign) NSUInteger uid;
+@property (nonatomic, strong) NSMutableSet<NSNumber *>* directions;
+
++ (instancetype)itemWithUid:(NSUInteger) uid;
+@end
+
+@implementation FBTVNavigationItem
+
++ (instancetype)itemWithUid:(NSUInteger) uid
+{
+  return [[FBTVNavigationItem alloc] initWithUid:uid];
+}
+
+- (instancetype)initWithUid:(NSUInteger) uid
+{
+  self = [super init];
+  if(self) {
+    _uid = uid;
+    _directions = [NSMutableSet set];
+  }
+  return self;
+}
+
+@end
+
+@interface FBTVNavigationTracker ()
+@property (nonatomic, strong) id<FBElement> targetElement;
+@property (nonatomic, assign) CGPoint targetCenter;
+@property (nonatomic, strong) NSMutableDictionary<NSNumber *, FBTVNavigationItem* >* navigationItems;
+@end
+
+@implementation FBTVNavigationTracker
+
++ (instancetype)trackerWithTargetElement:(id<FBElement>)targetElement
+{
+  FBTVNavigationTracker *tracker = [[FBTVNavigationTracker alloc] initWithTargetElement:targetElement];
+  tracker.targetElement = targetElement;
+  return tracker;
+}
+
+- (instancetype)initWithTargetElement:(id<FBElement>)targetElement
+{
+  self = [super init];
+  if(self) {
+    _targetElement = targetElement;
+    _targetCenter = FBRectGetCenter(targetElement.wdFrame);
+    _navigationItems = [NSMutableDictionary dictionary];
+  }
+  return self;
+}
+
+- (FBTVDirection)directionToMoveFocuse
+{
+  id<FBElement> focused = self.focusedElement;
```

Comment:
FBElement will change to `XCUIElement` in the next PR like https://github.com/KazuCocoa/WebDriverAgent/pull/4/files#diff-ca69d06a77426842d4a7b3c34bfcedc4R44